beps/0003-auth-architecture-evolution/README.md

```diff
@@ -68,9 +68,9 @@ Two new backend service interfaces are introduced to support these new features.
 
 The proposed design leaves the decision for how different endpoints are protected to the implementation of the plugin backends themselves. This includes whether particular routes should allow anonymous access, access from users authenticated via a cookie, or perhaps only allow access from other plugin backends and external services. This means that integrators do not need to - and do not have the ability to - configure access controls of individual endpoints, except for what the permission system already provides, and what is made available through static configuration or extension points.
 
-In order to allow for cookie-based authentication of incoming user requests, the `auth` plugin backend is extended to be able to issue user tokens with reduced scope, which in turn integrate with the new `AuthService` and `HttpAuthService`. The ability to use cookie auth for requests is an opt-in per route and is only be permitted for read methods (`GET`, `HEAD`, `OPTIONS`). The actual implementation of cookie-based flows will be up to each plugin, but with significant help from the new auth service interfaces.
+To ensure a secure-by-default design, there is a default access control policy that applies to all plugin routes, known as the "default auth policy". This policy is to only allow access from authenticated users and services, and is implemented in the `HttpRouterService` interface. In order to allow either unauthenticated access or cookie-based access, a plugin must opt-out of the default auth policy for specific path prefixes, effectively leaving the access control implementation to the plugin itself. This is done through the new `addAuthPolicy` method that is added to the `HttpRouterService` interface.
 
-In order to allow either unauthenticated access or cookie-based access, a plugin must explicitly opt-in the specific path prefixes that these should be available at. This is done through a new method that is added to the `HttpRouterService` interface.
+In order to allow for cookie-based authentication of incoming user requests, the `AuthService` is able to issue user tokens with limited scope. These limited scope tokens can still be used to fetch user information and in on-behalf-of service calls, but they are rejected by the default auth policy. The ability to use cookie auth for requests is an advanced use-case where plugins need to handle a lot of the auth logic, for example cookie storage and retrieval. These tokens with limited scope can also be user in other contexts where it is beneficial to avoid storing full user credentials, but instead use credentials that can be upgraded in a controlled manner, such as scaffolder tasks. The `AuthService` implementation can choose to have a longer expiry of the limited tokens compared to the full user tokens, but this is not a requirement.
```

Comment:
Thoughts on tagging these limited access tokens with a target path or endpoint? Seems a little odd that limited access tokens could be used anywhere across the application that accepts them. Something like S3 signed URLs is kind of what I'm thinking, restricted access to specific objects. The full flow is overkill for our use case, but the idea of having specific access to a bucket or file seems appealing. Though depending on how much cookie auth we do, I could see multiple separate cookies being an issue for header size.

Reply:
Yep kinda aiming to re-use cookies across plugins if possible, not a bad idea though. Not sure if it would really add that much in practice though. Idea behind the limited access tokens is that what you can actually do with them is strictly limited by backend plugins, where individual plugins are able to elevate the access for specific tasks. For example in the TechDocs backend you can use them to access doc pages, in the auth backend you can fetch user info, and in the scaffolder backend individual actions can use them in their service calls but a user is not able to start a scaffolding task with them.

We've discussed scoped access tokens at some length, but in the end we've mostly ended up making sure that the full users tokens don't get forwarded in service-to-service calls, using an OBO mechanism that reduces the scope of the token to a specific service call instead.

I think the inverse of this can still be worth exploring though, i.e. #972

Reply:
Yep could do something along those lines tbh. Funny thing is we have all of the information we need to be able to implement that afaik 😁, this is a plugin scoped service so we can make sure the cookie issuing and consumption are both scoped to the plugin in the future if we want to improve the system in that way. If you want limited tokens to leave the plugin you have to OBO anyway, so that should not be breaking